Lib/glyphsLib/glyphdata.py

```diff
@@ -31,69 +31,224 @@
+    # look up the individual parts.
+    base_name_parts = base_name.split("_")
+
+    # If all parts are in the AGLFN list, the glyph name is our production
+    # name already.
+    if all(part in agl.AGL2UV for part in base_name_parts):
+        if len(glyph_name) > 31:
+            return None
+        return glyph_name
+
+    _character_outside_BMP = False
+    production_names = []
+    for part in base_name_parts:
+        if part in agl.AGL2UV:
+            production_names.append(part)
+        else:
+            part_production_name = data.PRODUCTION_NAMES.get(part)
+            if part_production_name:
+                production_names.append(part_production_name)
+
+                # Note if there are any characters outside the Unicode BMP, e.g.
+                # "u10FFF" or "u10FFFF". Do not catch e.g. "u013B" though.
+                if len(part_production_name) > 5 and is_unicode_u_value(
+                    part_production_name
+                ):
+                    _character_outside_BMP = True
+
+            else:
+                return None
+
+    # Some names Glyphs uses resolve to other names that are not uniXXXX names and may
+    # contain dots (e.g. idotaccent -> i.loclTRK). If there is any name with a "." in
+    # it before the last element, punt. We'd have to introduce a "." into the ligature
+    # midway, which is invalid according to the AGL. Example: "a_i.loclTRK" is valid,
+    # but "a_i.loclTRK_a" isn't.
+    if any("." in part for part in production_names[:-1]):
+        return None
+
+    # If any production name starts with a "uni" there are none of the "uXXXXX" format,
+    # try to turn all parts into "uni" names and concatenate them.
+    if not _character_outside_BMP and any(
+        part.startswith("uni") for part in production_names
+    ):
+        uni_names = []
+
+        for part in production_names:
+            if part.startswith("uni"):
+                uni_names.append(part[3:])
+            elif len(part) == 5 and is_unicode_u_value(part):
+                uni_names.append(part[1:])
+            elif part in agl.AGL2UV:
+                uni_names.append("{:04X}".format(agl.AGL2UV[part]))
+            else:
+                return None
+
+        final_production_name = "uni{}".format("".join(uni_names)) + dot + suffix
     else:
-        unistr_result = None
-    category, subCategory = _get_category(name, unistr, data)
-    return Glyph(name, prodname, unistr_result, category, subCategory)
-
-
-def _get_unicode_category(unistr):
-    # We use data for a fixed Unicode version (3.2) so that our generated
-    # data files are independent of Python runtime that runs the rules.
-    # By switching to current Unicode data, we could save some entries
-    # in our exception tables, but the gains are not very large; only
-    # about one thousand entries.
-    if not unistr:
+        final_production_name = "{}".format("_".join(production_names)) + dot + suffix
```

Comment:
the `"{}".format(...)` here seems redundant, no? The `"_".join(production_names)` already returns a `str`.